AP_Math: added comments and a test for euler ordering

tridge opened the pull request:
our main euler functions did not have a comment on the ordering convention

peterbarker commented on libraries/AP_Math/matrix3.h:
```diff
@@ -235,13 +235,17 @@ class Matrix3 {
 
-    // create a rotation matrix from Euler angles
+    /*
+      create a rotation matrix from Euler angles in 321 euler orderin
```
```suggestion
      create a rotation matrix from Euler angles in 321 euler ordering
```

tridge replied:
changed

andyp1per commented:
Having scratched my head against this for some time it would be great to see the additional clarifying commentary:

- Say whether rotations are intrinsic or extrinsic? (https://github.com/ArduPilot/ardupilot/pull/27335)
- Say what the positive direction is for each axis
- This doc in matrix3.cpp now gives a 404 - http://gentlenav.googlecode.com/files/EulerAngles.pdf

andyp1per reviewed (commented):
I think its great as far as it goes, but possible could do with some additional commentary

isso11 commented:
> * positive direction is for each axis
I think so too. In the context of Euler angles and the North-East-Down (NED) coordinates frame:
Yaw (ψ): Positive is clockwise when viewed from above, around the Down (Z) axis.
Pitch (θ): Positive is nose up, around the East (Y) axis.
Roll (𝜙): Positive is right wing down, around the North (X) axis.

andyp1per commented:
@tridge I manually checked one of these and either euler 321 means something different to what I think it does or these rotations are euler 123. An easy example take ROTATION_ROLL_90_YAW_90, this means roll 90 followed by yaw 90. The code swaps around vector values so that [x, y, z] -> [z, x, y] which indeed is roll 90 followed by yaw 90 it is *not* yaw 90 followed by roll 90. So either the meaning of euler 321 is very weird - which I am prepared to believe - or your documentation is actually wrong.

andyp1per reviewed (changes requested):
Need to resolve my very confused understanding of the meaning of euler 321

isso11 commented:
> euler 321

Aren't these rotations basically for transformations between body fixed and local inertial NED frames? If so, then the rotations 321 should, by standard definition, mean  [Yaw (3, around the 3rd axis, Z), Pitch (2, around the 2nd axis, Y), Roll(1, around the 1st axis, X)] in that order.

andyp1per commented:
> Aren't these rotations basically for transformations between body fixed and local inertial NED frames? If so, then the rotations 321 should, by standard definition, mean [Yaw (3, around the 3rd axis, Z), Pitch (2, around the 2nd axis, Y), Roll(1, around the 1st axis, X)] in that order.

That is my understanding

Georacer commented:
> That is my understanding

Mine as well.

tridge commented:
@andyp1per @Georacer see the rotate_ordered() function in this PR

tridge commented:
> * Say whether rotations are intrinsic or extrinsic?

as demonstrated by the rotate_ordered() function it is intrinsic

tridge commented:
> * This doc in matrix3.cpp now gives a 404

it is common that URLs like this disappear. If we can find the PDF we could put it in https://github.com/ArduPilot/Datasheets

andyp1per commented:
@tridge I have pushed a test that fails to prove my point.

tridge commented:
@andyp1per I've found the original EulerAngles.pdf from Bill Premerlani and added it to out Datasheets repo and updated the link in the code in this PR

peterbarker commented:
Ping @andyp1per can we merge this one?

peterbarker commented:
Merged, thanks!
